segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,384 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.impl.extendeds3;
+
+import com.emc.object.Range;
+import com.emc.object.s3.S3Client;
+import com.emc.object.s3.S3Exception;
+import com.emc.object.s3.S3ObjectMetadata;
+import com.emc.object.s3.bean.AccessControlList;
+import com.emc.object.s3.bean.CanonicalUser;
+import com.emc.object.s3.bean.CopyPartResult;
+import com.emc.object.s3.bean.Grant;
+import com.emc.object.s3.bean.ListObjectsResult;
+import com.emc.object.s3.bean.MultipartPartETag;
+import com.emc.object.s3.bean.Permission;
+import com.emc.object.s3.request.CompleteMultipartUploadRequest;
+import com.emc.object.s3.request.CopyPartRequest;
+import com.emc.object.s3.request.PutObjectRequest;
+import com.emc.object.s3.request.SetObjectAclRequest;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import io.pravega.common.LoggerHelpers;
+import io.pravega.common.io.StreamHelpers;
+import io.pravega.common.util.ImmutableDate;
+import io.pravega.segmentstore.contracts.BadOffsetException;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentInformation;
+import io.pravega.segmentstore.contracts.StreamSegmentNotExistsException;
+import io.pravega.segmentstore.contracts.StreamSegmentSealedException;
+import io.pravega.segmentstore.storage.SegmentHandle;
+import io.pravega.segmentstore.storage.impl.filesystem.IdempotentStorageBase;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.AccessDeniedException;
+import java.time.Duration;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Storage adapter for extended S3 based storage.
+ *
+ * Each segment is represented as a single Object on the underlying storage.
+ *
+ * This implementation works under the assumption that data is only appended and never modified.
+ * Each block of data has an offset assigned to it and Pravega always writes the same data to the same offset. As a result
```

Comment:
Related to this, you may want to describe how we guarantee that the initial offset of a block is always the same. That's what gives the guarantee, the initial offset of a block is the same, the length of the block is the same, and the bytes are the same.